R/finalize_plot.R

```diff
@@ -216,59 +230,167 @@ finalize_plot <- function(plot = NULL,
-        set_default_aes(geom_defaults)
-      }
-      stop("An error occurred in ggplot preparation", call. = FALSE)
-    }
+  #  top line
+  grobs$topline <- grid::linesGrob(
+    name = "topline",
+    default.units = "bigpts",
+    x = c(0, consts$width),
+    y = consts$height - consts$margin_topline_t,
+    gp = grid::gpar(col = cmapplot_globals$colors$blackish,
+                    lineend = "butt",
+                    lwd = consts$lwd_topline / .lwd)
   )
 
-  # reset geom defaults
-  if (use_cmap_aes) {
-    set_default_aes(geom_defaults)
+  # in traditional (horizontal) mode, create title and left-side caption boxes
+  if(!vert_mode){
+
+    # title textbox
+    grobs$title <- gridtext::textbox_grob(
+      name = "title",
+      text = title,
+      default.units = "bigpts",
+      # set location down from top left corner
+      x = 0,
+      y = consts$height - consts$margin_title_to_top,
+      hjust = 0,
+      vjust = 1,
+      # set dimensions
+      width = consts$title_width,
+      maxheight = consts$height - consts$margin_title_to_top,
+      # retract texbox size on left
+      margin = grid::unit(c(0, 0, # top, right
+                            consts$margin_title_b, # bottom
+                            consts$margin_title_l), # left
+                          "bigpts"),
+      # set font aesthetic variables
+      gp = grid::gpar(fontsize=cmapplot_globals$fsize$L,
+                      fontfamily=cmapplot_globals$font$strong$family,
+                      fontface=cmapplot_globals$font$strong$face,
+                      lineheight=consts$leading_title,
+                      col=cmapplot_globals$colors$blackish),
+      box_gp = grid::gpar(col = ifelse(debug, "red", NA),
+                          fill = NA)
+    )
+
+    # caption textbox
+    grobs$caption <- gridtext::textbox_grob(
+      name = "caption",
+      text = caption,
+      default.units = "bigpts",
+      # set location
+      x = 0,
+      y = 0,
+      hjust = 0,
+      vjust = 0,
+      # set dimensions
+      width = consts$title_width,
+      height = consts$height - consts$margin_title_to_top - safe_grobHeight(grobs$title),
+      # retract texbox size on each side
+      margin = grid::unit(c(0, 0,  # top, right
+                            consts$margin_caption_b,# bottom
+                            consts$margin_title_l), # left
+                          "bigpts"),
+      # set aesthetic variables
+      valign = caption_align,
+      gp = grid::gpar(fontsize = cmapplot_globals$fsize$S,
+                      fontfamily = cmapplot_globals$font$light$family,
+                      fontface = cmapplot_globals$font$light$face,
+                      lineheight = consts$leading_caption,
+                      col = cmapplot_globals$colors$blackish),
+      box_gp = grid::gpar(col = ifelse(debug, "red", NA),
+                          fill = NA)
+    )
+  }
+
+  # in vertical mode, and if caption exists, create bottom caption box (but no title)
```

Comment:
I understand the rationale for this - however, I think that it may be helpful in the future to have the ability to have a title in the vertical layout, potentially above the bar (which was what I had implemented, however jankily, in #109 ). It would create some more complexity, because then you need to have an adjustment to the top line positioning - although I think we could accomplish that with a similar use of the safe_grobheight function you've developed (that one is nice, by the way). Not an urgent function, but something to think about for the future.

Reply:
Agreed. With `safe_grobHeight()` and the new conditional grob build framework, it shouldn't be hard to install an alternate title grob above the topline. If you wanted to put together a PR building off this code, I don't see a reason not to implement that too.

Reply:
Agreed. My apologies for missing that part of your work in #109. With `safe_grobHeight()` and the new conditional grob build framework, it shouldn't be hard to install an alternate title grob above the topline. It would be really easy, I think... if a new grob top_title (or maybe we change to `title_vert` and `caption_vert`) gets created before the topline does, the topline height could be set to the top margin plus `safe_grobheight(top_title)`...

If you wanted to put together a PR that (re)implements this, I don't see a reason not to implement it.